de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidatesRestControllerImpl.java

```diff
@@ -0,0 +1,135 @@
+import org.springframework.context.annotation.Profile;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.google.common.collect.ImmutableList;
+
+import de.metas.Profiles;
+import de.metas.invoicecandidate.api.IInvoiceCandBL;
+import de.metas.invoicecandidate.api.IInvoiceCandidateEnqueueResult;
+import de.metas.invoicecandidate.model.I_C_Invoice_Candidate;
+import de.metas.logging.LogManager;
+import de.metas.process.IADPInstanceDAO;
+import de.metas.process.PInstanceId;
+import de.metas.rest_api.invoicecandidates.InvoiceCandidatesRestEndpoint;
+import de.metas.rest_api.invoicecandidates.request.JsonInvoiceCandCreateRequest;
+import de.metas.rest_api.invoicecandidates.request.JsonInvoiceCandidate;
+import de.metas.rest_api.invoicecandidates.response.JsonInvoiceCandCreateResponse;
+import de.metas.rest_api.utils.JsonErrors;
+import de.metas.util.Services;
+import de.metas.util.rest.ExternalId;
+import lombok.NonNull;
+
+/*
+ * #%L
+ * de.metas.business.rest-api-impl
+ * %%
+ * Copyright (C) 2019 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ * Used for managing invoices and invoice candidates(create, query)
+ */
+@RestController
+@RequestMapping(InvoiceCandidatesRestEndpoint.ENDPOINT)
+@Profile(Profiles.PROFILE_App)
+class InvoiceCandidatesRestControllerImpl implements InvoiceCandidatesRestEndpoint {
+
+	private static final Logger logger = LogManager.getLogger(InvoiceCandidatesRestControllerImpl.class);
+
+	private final IADPInstanceDAO adPInstanceDAO = Services.get(IADPInstanceDAO.class);
+	private final transient IInvoiceCandBL invoiceCandBL = Services.get(IInvoiceCandBL.class);
+	private final InvoiceJsonConverters jsonConverters;
+
+	public InvoiceCandidatesRestControllerImpl(@NonNull final InvoiceJsonConverters jsonConverters) {
+		this.jsonConverters = jsonConverters;
+	}
+
+	@PostMapping(consumes = { "application/json" })
+	@Override
+	public ResponseEntity<JsonInvoiceCandCreateResponse> createInvoices(
+			@RequestBody @NonNull final JsonInvoiceCandCreateRequest request) {
+		try {
```

Comment:
please extract all this query-code into nother class, to keep this endpoint clean.

Reply:
Good point you wrote this as I wanted to ask you where it should be better to extract this? In a class in the same de.metas.business.rest-api-impl project or in another one? I don't know exactly.

Reply:
for now, please extract it to a service inside de.metas.business.rest-api-impl project, near this endpoint implementation.
we can later consider to move it to the "core", but then we'd also need some in-between-code

Reply:
done.

Reply:
PSS: pls coordinate with @metas-ts, maybe @metas-ts wants to handle the suggestion above

Reply:
Totally agree. Will do.